src/config.js

```diff
@@ -0,0 +1,62 @@
+/* @flow */
+import path from 'path';
+
+import requireUncached from 'require-uncached';
+
+import {createLogger} from './util/logger';
+import {UsageError} from './errors';
+
+const log = createLogger(__filename);
+
+type ApplyConfigToArgvParams = {|
+  argv: Object,
+  configObject: Object,
+  defaultValues: Object,
+|};
+
+type ParseConfigParams = {|
+  sourceDir: string,
+  configFileName: string,
+|};
+
+export function applyConfigToArgv({
+  argv,
+  configObject,
+  defaultValues = {},
+}: ApplyConfigToArgvParams): Object {
+  for (const option in configObject) {
+    if (argv[option] && defaultValues[option]) {
+      const wasValueSetOnCLI = argv[option] !== defaultValues[option];
+      if (wasValueSetOnCLI) {
```

Comment:
Add some debug logging here so we can detect when this happens (since we know there are subtle edge cases when it will be wrong). I suggest:
````js
log.debug(`Favoring CLI ${option}=${argv[option]} over configuration: ${option}=${configObject[option]}`);
````